File: Prefix Sum/Python/Product array puzzle.py
#User function Template for python3

class Solution:
    def productExceptSelf(self, arr):
        #code here
        zero=0
        idx=-1
        pro=1
        
        for i in range(len(arr)):
            if arr[i]==0:
                zero+=1
                idx=i
            else:
                pro*=arr[i]
        res=[0]*len(arr)
        if zero==0:
            for i in range(len(arr)):
                res[i]=pro//arr[i]
        elif zero==1:
            res[idx]=pro
        return res

File: Prefix Sum/C++/Product array puzzle.cpp
class Solution {
  public:
    vector<int> productExceptSelf(vector<int>& arr) {
        // code here
        int n = arr.size();
        vector<int> preProduct(n), sufProduct(n), res(n);
        preProduct[0] = 1;
        
        for (int i = 1; i < n; i++) {
            preProduct[i] = arr[i - 1] * preProduct[i - 1];
        }
        
        sufProduct[n - 1] = 1;
        for (int j = n - 2; j >= 0; j--) {
            sufProduct[j] = arr[j + 1] * sufProduct[j + 1];
        }
        
        for (int i = 0; i < n; i++) {
            res[i] = preProduct[i] * sufProduct[i];
        }
        return res;
    }
};
